Nmap os detection (#1349)
Codecademy/docs: content/cybersecurity/concepts/nmap/terms/os-detection/os-detection.md

---
Title: 'OS Detection'

> Description: 'One of the best-known features of Nmap is remote OS detection using TCP/IP stack fingerprinting.'

Reviewer: ```suggestion
Description: 'Remotely scans a target host and presents details of its operating system if there is a match.'
```

For term entries in Docs, let's set up the `Description` such that it begins with a present-tense verb 😄

Subjects:
  - 'Computer Science'
  - 'Networking'
Tags:
  - 'Networking'
  - 'Cybersecurity'
  - 'Ethical Hacking'
CatalogContent:
  - 'introduction-to-cybersecurity'
  - 'paths/scan-systems-with-nmap'
SeeAlso:
  - 'aggressive-scan'
---

One of Nmap's best known features is remote **OS detection** using TCP/IP stack fingerprinting. With OS detection, Nmap sends a series of TCP and UDP packets to the specified host and examines the responses. Nmap then compares the results to its *nmap-os-db* database of more than 2,600 known OS fingerprints and prints out the OS details if there is a match.  

> Each fingerprint includes a freeform textual description of the OS, and a classification which provides the vendor name (e.g. Sun), underlying OS (e.g. Solaris), OS generation (e.g. 10), and device type (general purpose, router, switch, game console, etc). Most fingerprints also have a Common Platform Enumeration (CPE) representation, like `cpe:/o:linux:linux_kernel:2.6.`

Author: I thought about this one for a few minutes, I eventually warmed up to the idea.
Lets run with it :smiley:

## Example

The following example runs a Nmap Scan and enables OS detection on the site [scanme.nmap.org](http://scanme.nmap.org/):

```shell
nmap -O scanme.nmap.org
```
> **Note:** This is a site explicitly set up for testing Nmap. Scans should only be performed on domains with explicit permissions given to the user.

This results in the following output:

```shell
Starting Nmap 7.93 ( https://nmap.org ) at 2022-12-17 19:59 EST
Nmap scan report for scanme.nmap.org (45.33.32.156)
Host is up (0.089s latency).
Other addresses for scanme.nmap.org (not scanned): 2600:3c01::f03c:91ff:fe18:bb2f
Not shown: 992 closed tcp ports (reset)
PORT      STATE    SERVICE
22/tcp    open     ssh
25/tcp    filtered smtp
80/tcp    open     http
135/tcp   filtered msrpc
139/tcp   filtered netbios-ssn
445/tcp   filtered microsoft-ds
9929/tcp  open     nping-echo
31337/tcp open     Elite
Device type: general purpose
Running: Linux 5.X
OS CPE: cpe:/o:linux:linux_kernel:5
OS details: Linux 5.0 - 5.4
Network Distance: 10 hops

OS detection performed. Please report any incorrect results at https://nmap.org/submit/ .
Nmap done: 1 IP address (1 host up) scanned in 6.96 seconds

> ```

Author: added suggested fingerprint info below ```shell snippet

---
Title: 'OS Detection'

> Description: 'One of Nmap's best-known features is remote OS detection using TCP/IP stack fingerprinting.'

Reviewer 2: ```suggestion
Description: 'One of the best-known features of Nmap is remote OS detection using TCP/IP stack fingerprinting.'
```

Subjects:
  - 'Computer Science'
  - 'Networking'
Tags:
  - 'Networking'
  - 'Cybersecurity'
  - 'Ethical Hacking'
CatalogContent:
  - 'introduction-to-cybersecurity'
  - 'paths/scan-systems-with-nmap'
SeeAlso:
  - 'aggressive-scan'
---

> One of Nmap's best known features is remote **OS Detection** using TCP/IP stack fingerprinting. With **OS Detection**, Nmap sends a series of TCP and UDP packets to the specified host and examine the reponces. Nmap then compares the results to its *nmap-os-db* database of more than 2,600 known OS fingerprints and prints out the OS details if there is a match.  

Reviewer 2: ```suggestion
One of Nmap's best known features is remote **OS detection** using TCP/IP stack fingerprinting. With OS detection, Nmap sends a series of TCP and UDP packets to the specified host and examines the responses. Nmap then compares the results to its *nmap-os-db* database of more than 2,600 known OS fingerprints and prints out the OS details if there is a match.  
```

> **Operating System (OS) detection** is a feature in Nmap that remotely [scans](https://www.codecademy.com/resources/docs/cybersecurity/nmap/nmap-scan) a target host and presents details of its operating system if there is a match. TCP/IP stack fingerprinting is used to send a series of probes (e.g., TCP and UDP packets) to the specified host and examines the responses. The results are then compared to the `nmap-os-db` database 2,600+ known OS fingerprints, and the details are printed if there is a match.

Reviewer: ```suggestion
**Operating system (OS) detection** is a feature in Nmap that remotely [scans](https://www.codecademy.com/resources/docs/cybersecurity/nmap/nmap-scan) a target host and presents details of its operating system if there is a match. TCP/IP stack fingerprinting is used to send a series of probes (e.g., TCP and UDP packets) to the specified host and examines the responses. The results are then compared to the `nmap-os-db` database 2,600+ known OS fingerprints, and the details are printed if there is a match.
```

> 
> 
> 
> > **Note:** Root privileges are required to properly perform the scan.
> 

Reviewer: ```suggestion

> **Note:** Root privileges are required to properly perform the scan.

```

> Note that *root* privilages are required to properly perform the scan.

Reviewer 2: ```suggestion
> **Note:** Root privileges are required to properly perform the scan.
```

> 

Reviewer 2: Eliminate extra line break.

Author: removed extra line break

> **OS Detection** is enabled with:

Reviewer 2: ```suggestion
OS detection is enabled with the `-O` option.
```

> - Enable OS detection (`-O`)

Reviewer 2: Delete this and include it in the line above, as I did with my suggestion.

> **OS Detection** is controlled with the following options (be sure to include with `-O`):

Reviewer 2: ```suggestion
OS detection is controlled with the following options (be sure to include with `-O`):
```

> - Limit **OS detection** to promising targets (`--osscan-limit`)

Reviewer 2: ```suggestion
- Limit OS detection to promising targets (`--osscan-limit`)
```

>     - This can save substantial time by only scanning targets that have a critera that makes **OS Detection** more effective, which is if there is at least one open and one closed TCP port are found. Nmap will now attempt **OS Detection** against hosts which do not meet that criteria.

Reviewer 2: ```suggestion
    - This can save substantial time by only scanning targets that have a criteria that makes OS detection more effective, which is if there is at least one open and one closed TCP port are found. Nmap will not attempt OS detection against hosts which do not meet that criteria.
```

> - Guess **OS detection** results (`--osscan-guess; --fuzzy`)

Reviewer 2: ```suggestion
- Guess OS detection results (`--osscan-guess; --fuzzy`)
```

> - `--osscan-limit` - Limits detection to only host targets with at least one open and one closed TCP port, and ignores hosts that do not meet this criteria.
> - `--osscan-guess; --fuzzy` - Presents possible results if a perfect OS match can't be found.
>   - The match has to be very close for Nmap to do this by default. Either of these (equivalent) options make Nmap guess more aggressively. Nmap will still tell when an imperfect match is printed and display its confidence level (0% - 100%) for each guess.
> - `--max-os-tries` - Sets the maximum number of detection attempts against a target.
>   - By default, Nmap tries five times if conditions are favorable or two times when conditions aren't so good. A specific amount of tries can be specified; the lower the amount, the quicker the scan.
> - Provides more information from an OS detection scan (`-v`), including a incremental ID for each IP packet header sent.

Reviewer: ```suggestion
- `--osscan-limit`: Limits detection to only host targets with at least one open and one closed TCP port, and ignores hosts that do not meet this criteria.
- `--osscan-guess; --fuzzy`: Presents possible results if a perfect OS match can't be found.
  - The match has to be very close for Nmap to do this by default. Either of these (equivalent) options make Nmap guess more aggressively. Nmap will still tell when an imperfect match is printed and display its confidence level (0% - 100%) for each guess.
- `--max-os-tries`: Sets the maximum number of detection attempts against a target.
  - By default, Nmap tries five times if conditions are favorable or two times when conditions aren't so good. A specific amount of tries can be specified; the lower the amount, the quicker the scan.
- `-v`: Provides more information from an OS detection scan, including an incremental ID for each IP packet header sent.
```

>     - When Nmap is unable to detect a perfect OS match, it sometimes offers up near-matches as possibilities. The match has to be very close for Nmap to do this by default. Either of these (equivalent) options make Nmap guess more aggressively. Nmap will still tell you when an imperfect match is printed and display its confidence level (percentage) for each guess.

Reviewer 2: ```suggestion
    - When Nmap is unable to detect a perfect OS match, it sometimes offers up near-matches as possibilities. The match has to be very close for Nmap to do this by default. Either of these (equivalent) options make Nmap guess more aggressively. Nmap will still tell when an imperfect match is printed and display its confidence level (percentage) for each guess.
```

> - Set the maximum number of **OS detection** tries against a target (`--max-os-tries`)

Reviewer 2: ```suggestion
- Set the maximum number of OS detection tries against a target (`--max-os-tries`)
```

> - Increase verbosity level to your **OS Detection** Scan (`-v`)

Reviewer 2: ```suggestion
- Increase verbosity level to OS detection Scan (`-v`)
```

> 

Reviewer 2: Eliminate extra line break

Author: Removed extra line break

## Syntax 

```pseudo
nmap -O <target>
```

> The above conducts an Nmap scan with **OS Detection** on `<target>`.

Reviewer 2: ```suggestion
The above conducts an Nmap scan with OS detection on `<target>`.
```

## Example

> The following example runs a Nmap Scan and enables **OS Detection** on the site [scanme.nmap.org](http://scanme.nmap.org/):

Reviewer 2: ```suggestion
The following example runs a Nmap Scan and enables OS detection on the site [scanme.nmap.org](http://scanme.nmap.org/):
```

This results in the following output:

```shell
Starting Nmap 7.93 ( https://nmap.org ) at 2022-12-17 19:59 EST
Nmap scan report for scanme.nmap.org (45.33.32.156)
Host is up (0.089s latency).
Other addresses for scanme.nmap.org (not scanned): 2600:3c01::f03c:91ff:fe18:bb2f
Not shown: 992 closed tcp ports (reset)
PORT      STATE    SERVICE
22/tcp    open     ssh
25/tcp    filtered smtp
80/tcp    open     http
135/tcp   filtered msrpc
139/tcp   filtered netbios-ssn
445/tcp   filtered microsoft-ds
9929/tcp  open     nping-echo
31337/tcp open     Elite
Device type: general purpose
Running: Linux 5.X
OS CPE: cpe:/o:linux:linux_kernel:5
OS details: Linux 5.0 - 5.4
Network Distance: 10 hops

OS detection performed. Please report any incorrect results at https://nmap.org/submit/ .
Nmap done: 1 IP address (1 host up) scanned in 6.96 seconds
```

> Each fingerprint includes a freeform textual description of the OS, and a classification which provides the following:

Reviewer: ```suggestion

Each fingerprint includes a freeform textual description of the OS, and a classification which provides the following:
```
